src/gateway.c

```diff
@@ -727,12 +745,27 @@ static void handle_exec_sql_cb(struct exec *exec, int status)
 	req->exec_count += 1;
 	sqlite3_finalize(exec->stmt);
 
-	if (status == SQLITE_DONE) {
-		handle_exec_sql_next(g, req, true);
-	} else {
+	if (status != SQLITE_DONE) {
 		assert(g->leader != NULL);
 		failure(req, status, error_message(g->leader->conn, status));
 		g->req = NULL;
+		return;
+	}
+
+	/* It would be valid to always invoke handle_exec_sql_next directly
+	 * here. But that can lead to bounded recursion when we have several
+	 * `;`-separated statements in a row that do not generate rows. To make
+	 * sure the stack depth stays under control, we defer have the event
+	 * loop invoke handle_exec_sql_next itself on the next iteration, but
+	 * only if there is a prior call to handle_exec_sql_next above us on
+	 * the stack. We also invoke handle_exec_sql_next directly if the
+	 * gateway doesn't have access to an event loop (this is only the case
+	 * in the unit tests). */
+	if (exec->async || g->defer.loop == NULL) {
+		handle_exec_sql_next(g, req, true);
```

Comment:
Question: Is it worth having an explicit recursion counter as a parameter of a function or an associated structure?
Just curious.

Reply:
Good question---I think the best version of this would be to have a depth counter field on `struct exec` and `struct barrier`. But then with the approach in this PR, that field would only ever be 0 or 1, so we'd just be treating it as a boolean anyway. (Unless we wanted to allow some positive recursion depth, but I'm not sure how to pick that constant.)